subsys/bluetooth/audio/bap_base.c

```diff
@@ -0,0 +1,572 @@
+#define BASE_SUBGROUP_MAX_COUNT (BASE_MAX_SIZE / BASE_SUBGROUP_MIN_SIZE)
+
+static uint32_t base_pull_pd(struct net_buf_simple *net_buf)
+{
+	return net_buf_simple_pull_le24(net_buf);
+}
+
+static uint8_t base_pull_bis_count(struct net_buf_simple *net_buf)
+{
+	return net_buf_simple_pull_u8(net_buf);
+}
+
+static void base_pull_codec_id(struct net_buf_simple *net_buf,
+			       struct bt_bap_base_codec_id *codec_id)
+{
+	struct bt_bap_base_codec_id codec;
+
+	codec.id = net_buf_simple_pull_u8(net_buf);    /* coding format */
+	codec.cid = net_buf_simple_pull_le16(net_buf); /* company id */
+	codec.vid = net_buf_simple_pull_le16(net_buf); /* VS codec id */
+
+	if (codec_id != NULL) {
+		*codec_id = codec;
+	}
+}
+
+static uint8_t base_pull_ltv(struct net_buf_simple *net_buf, uint8_t **data)
+{
+	const uint8_t len = net_buf_simple_pull_u8(net_buf);
+
+	if (data == NULL) {
+		net_buf_simple_pull_mem(net_buf, len);
+	} else {
+		*data = net_buf_simple_pull_mem(net_buf, len);
+	}
+
+	return len;
+}
+
+const struct bt_bap_base *bt_bap_base_get_base_from_ad(const struct bt_data *ad)
+{
+	struct bt_uuid_16 broadcast_uuid;
+	const struct bt_bap_base *base;
+	struct net_buf_simple net_buf;
+	uint8_t subgroup_count;
+	void *uuid;
+
+	CHECKIF(ad == NULL) {
+		LOG_DBG("data is NULL");
+
+		return NULL;
+	}
+
+	if (ad->type != BT_DATA_SVC_DATA16) {
+		LOG_DBG("Invalid type: %u", ad->type);
+
+		return NULL;
+	}
+
+	if (ad->data_len < BASE_MIN_SIZE) {
+		LOG_DBG("Invalid len: %u", ad->data_len);
+
+		return NULL;
+	}
+
+	net_buf_simple_init_with_data(&net_buf, (void *)ad->data, ad->data_len);
+
+	uuid = net_buf_simple_pull_mem(&net_buf, BT_UUID_SIZE_16);
+	if (!bt_uuid_create(&broadcast_uuid.uuid, uuid, BT_UUID_SIZE_16)) {
+		LOG_ERR("bt_uuid_create failed");
+
+		return NULL;
+	}
+
+	if (bt_uuid_cmp(&broadcast_uuid.uuid, BT_UUID_BASIC_AUDIO) != 0) {
+		LOG_DBG("Invalid UUID");
+
+		return NULL;
+	}
+
+	/* Store the start of the BASE */
+	base = (const struct bt_bap_base *)net_buf.data;
+
+	/* Pull all data to verify that the result BASE is valid */
+	base_pull_pd(&net_buf);
+	subgroup_count = net_buf_simple_pull_u8(&net_buf);
+	if (subgroup_count == 0 || subgroup_count > BASE_SUBGROUP_MAX_COUNT) {
+		LOG_DBG("Invalid subgroup count: %u", subgroup_count);
+
+		return NULL;
+	}
+
+	for (uint8_t i = 0U; i < subgroup_count; i++) {
+		uint8_t bis_count;
+		uint8_t ltv_len;
+
+		if (net_buf.len < sizeof(bis_count)) {
+			LOG_DBG("Invalid BASE length: %u", ad->data_len);
+
+			return NULL;
+		}
+
+		bis_count = base_pull_bis_count(&net_buf);
+		if (bis_count == 0 || bis_count > BT_ISO_MAX_GROUP_ISO_COUNT) {
+			LOG_DBG("Subgroup[%u]: Invalid BIS count: %u", i, bis_count);
+
+			return NULL;
+		}
+
+		if (net_buf.len < BASE_CODEC_ID_SIZE) {
+			LOG_DBG("Invalid BASE length: %u", ad->data_len);
+
+			return NULL;
+		}
+
+		base_pull_codec_id(&net_buf, NULL);
+
+		/* Pull CC */
+		if (net_buf.len < sizeof(ltv_len)) {
```

Comment:
Could lines 153-165 be moved into a function, since reading ltv data is done 3 times?